File: SecureBlackbox Samples/DTLS Server/cpp/dtlsserver.cpp
/*
 * SecureBlackbox 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <ctype.h>
#include <fstream>
#include <stdlib.h>
#include "../../include/secureblackbox.h"
#include <cstring>

#define LINE_LEN 80

class MyDTLSServer : public DTLSServer {
    int FireTLSShutdown(DTLSServerTLSShutdownEventParams* e) override {
        printf(" [%lld] Secure session closed\n", e->ConnectionID);
        return 0;
    };

    int FireTLSEstablished(DTLSServerTLSEstablishedEventParams* e) override {
        printf(" [%lld] Secure session established\n", e->ConnectionID);
        return 0;
    }

    int FireError(DTLSServerErrorEventParams* e) override {
        printf(" [%lld] Error %i: %s\n", e->ConnectionID, e->ErrorCode, e->Description);
        return 0;
    }

    int FireData(DTLSServerDataEventParams* e) override {
        const int len = e->lenBuffer;
        char* bufstr = new char[len + 1];
        strncpy(bufstr, e->Buffer, len); // get string from buffer
        bufstr[len] = '\0';
        printf("[%lld] C->S: %s\n", e->ConnectionID, bufstr);
        char* rev = new char[len + 1];
        strcpy(rev, bufstr);
        for (int i = 0, j = len - 1; i < j; i++, j--) {
            std::swap(rev[i], rev[j]);
        }
        SendText(e->ConnectionID, rev);
        printf("[%lld] S->C: %s\n", e->ConnectionID, rev);
        delete[] rev;
        delete[] bufstr;
        return 0;
    }

    int FireAccept(DTLSServerAcceptEventParams* e) override {
        e->Accept = true;
        printf("Accepted a new client from %s:%i\n", e->RemoteAddress, e->RemotePort);
        return 0;
    }
};

int main(int argc, char* argv[]) {

    if (argc < 2) {
        fprintf(stderr, "\n");
        fprintf(stderr, "  port         the port of server\n");
        fprintf(stderr, "  filename     the certificate path\n");
        fprintf(stderr, "  password     the certificate password\n");
        fprintf(stderr, "  \nExample    dtlsserver port filename.cert\n");
        fprintf(stderr, "  \nExample    dtlsserver port filename.cert \"123456\"\n");
        printf("Press enter to continue.\n");
        getchar();
    } else {
        MyDTLSServer server;
        CertificateManager certManager;
        int ret_code = 0;

        char command[LINE_LEN]; // user's command
        char msg[10000]; // storage for mail message

        if (server.GetActive()) {
            printf("Server already started\n");
            goto done;
        }
        if (argc > 2) {
            ret_code = certManager.ImportFromFile(argv[2], argc > 3 ? argv[3] : "");
        } else {
            printf("A server certificate is required for this demo.\n");
            goto done;
        }

        if (certManager.GetLastErrorCode() == 0) {
            server.SetServerCertCount(server.GetServerCertCount() + 1);
            server.SetServerCertHandle(server.GetServerCertCount() - 1, certManager.GetCertHandle());
            printf("Server certificate loaded\n");
        } else {
            printf("Failed to load certificate.\n");
            goto done;
        }
        for (int i = 0; i < strlen(argv[1]); i++) {
            if (!isdigit(argv[1][i])) {
                printf("Wrong port.\n");
                goto done;
            }
        }
        server.SetPort(atoi(argv[1]));

        if (ret_code = server.Start()) {
            goto done;
        }

        printf("The server started successfully!\n");
        printf("Type \"stop\" to stop the server.\n\n");

        while (1) {
            strcpy(msg, "");
            printf("> ");
            fgets(command, LINE_LEN, stdin);
            command[max(0, strlen(command) - 1)] = '\0';
            for (int i = 0; i < strlen(command); i++) {
                command[i] = tolower(command[i]);
            }

            if (!strcmp(command, "?")) {
                printf("  stop          stop server\n");
            } else if (!strcmp(command, "stop")) {
                if (server.GetActive()) {
                    ret_code = server.Stop();
                    if (ret_code == 0) {
                        printf("The server was stopped!");
                    } else {
                        goto done;
                    }
                }
                exit(0);
            } else {
                printf("Bad command!\n");
            }
        }
    done:
        if (ret_code) // Got an error.  The user is done.
        {
            if (server.GetActive()) {
                server.Stop();
            }
            printf("\nError: %d\n", ret_code);
            if (server.GetLastError()) {
                printf(" \"%s\"\n", server.GetLastError());
            }
        }
        printf("Exiting... (press enter)\n");
        getchar();
        exit(ret_code);
    }
}

File: SecureBlackbox Samples/DTLS Server/net/dtlsserver.cs
/*
 * SecureBlackbox 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of SecureBlackbox in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/secureblackbox
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using System.Text;
using nsoftware.SecureBlackbox;
using SBMath;

class dtlsserver
{
    private static string optval(string[] args, string option)
    {
        for (int x = 0; x < args.Length - 1; x++)
        {
            if (args[x].Equals(option, StringComparison.CurrentCultureIgnoreCase))
            {
                return args[x + 1];
            }
        }
        return "";
    }

    private static bool optext(string[] args, string option)
    {
        for (int x = 0; x < args.Length; x++)
        {
            if (args[x].Equals(option, StringComparison.CurrentCultureIgnoreCase))
            {
                return true;
            }
        }
        return false;
    }

    private static void displayHelp(string errMes)
    {
        Console.WriteLine(
                "NAME\n" +
                "  dtlsserver -- SecureBlackbox DTLSServer Demo Application\n\n" +
                "SYNOPSIS\n" +
                "  dtlsserver <port> <-cert certificate_file> [-certpass certificate_password]\n\n" +
                "DESCRIPTION\n" +
                "  This sample is a very simple echo server that shows how to use the DTLSServer component\n" +
                "  to receive messages from DTLS clients and send them back to the clients.\n\n" +
                "  The options are as follows:\n\n" +
                "  port         The port of server (Required).\n\n" +
                "  -cert        The certificate used to sign files (Required).\n\n" +
                "  -certpass    The password for the signing certificate.\n\n" +
                "EXAMPLES\n" +
                "  dtlsserver 4433 -cert C:\\certs\\mycert.pfx -certpass mypassword\n"
        );

        if (errMes.Length > 0)
        {
            Console.WriteLine("Error: " + errMes);
            Console.WriteLine();
        }

        confirmExit();
    }

    private static void confirmExit()
    {
        Console.WriteLine("Press Enter to exit the demo.");
        Console.ReadLine();
    }

    private static DTLSServer server;

    private static void _server_OnAccept(object sender, DTLSServerAcceptEventArgs e)
    {
        e.Accept = true;
        Console.WriteLine("Accepted a new client from " + e.RemoteAddress + ":" + e.RemotePort);
    }

    private static void _server_OnData(object sender, DTLSServerDataEventArgs e)
    {
        Console.WriteLine(" [" + e.ConnectionID + "] C->S: " + Encoding.UTF8.GetString(e.Buffer));

        byte[] dst = e.Buffer;
        Array.Reverse(dst);

        try
        {
            server.SendData(e.ConnectionID, dst);
            Console.WriteLine(" [" + e.ConnectionID + "] S->C: " + Encoding.UTF8.GetString(dst));
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error: " + ex.Message);
        }
    }

    private static void _server_OnError(object sender, DTLSServerErrorEventArgs e)
    {
        Console.WriteLine(" [" + e.ConnectionID + "] Error " + e.ErrorCode + ": " + e.Description);
    }

    private static void _server_OnTLSEstablished(object sender, DTLSServerTLSEstablishedEventArgs e)
    {
        Console.WriteLine(" [" + e.ConnectionID + "] Secure session established");
    }

    private static void _server_OnTLSShutdown(object sender, DTLSServerTLSShutdownEventArgs e)
    {
        Console.WriteLine(" [" + e.ConnectionID + "] Secure session closed");
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            displayHelp("port is required.");
            return;
        }

        server = new DTLSServer();
        try
        {
            server.Port = int.Parse(args[0]);
        }
        catch (Exception ex)
        {
            displayHelp("Invalid port value");
            return;
        }

        try
        {
            server.OnAccept += _server_OnAccept;
            server.OnData += _server_OnData;
            server.OnError += _server_OnError;
            server.OnTLSEstablished += _server_OnTLSEstablished;
            server.OnTLSShutdown += _server_OnTLSShutdown;

            if (optext(args, "-cert"))
            {
                CertificateManager cm = new CertificateManager();
                cm.ImportFromFile(optval(args, "-cert"), optval(args, "-certpass"));
                server.TLSServerChain.Add(cm.Certificate);
            }
            else
            {
                displayHelp("-cert is required.");
                return;
            }

            server.Start();

            Console.WriteLine("DTLS server started on port " + server.Port + ". Press enter to stop server");
            Console.ReadLine();

            server.Stop();

            Console.WriteLine("Server stopped.\n");

            confirmExit();
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error: " + ex.Message);
        }
    }
}





class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add a key to the dictionary for each argument.
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/", then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}